rfcs/0089-spill-hash-aggregation.md

```diff
@@ -0,0 +1,98 @@
+
+How to choose a partition number? If we have the priori information about the aggregation input size, we could choose a proper partition number to ensure each partition could be fitted into memory, however, risingwave doesn't contain this kind of statistic, so we need to choose a partition carefully.
+Choosing too few partitions leads to a handful of large-sized partitions causing extra rounds of Spill and a large amount of spilling to disk. On the other hand, while using a larger number of partitions can reduce the total amount of spilling, it can make the agg's I/O pattern more random due to frequent writings of partitions containing just a few rows. According to this paper[1], we can use 20 as a default partition number.
+
+```
+  HashAggExec.execute:
+
+    // process hash agg in memory
+    ...
+
+    if not enough memory:
+      // Spill
+      for (k,v) in hash_table:
+        partition = hash(k) % partition_num
+        write_to_agg_state_file(k, v)
+      for chunk in inputs:
+        partition = hash(k) % partition_num
+        write_to_input_file(chunk)
+
+      // process partition by partition
+      for i in partition_num:
+        HashAggExec.execute(agg_state_file[i].read(), input_file[i].read())
+
+```
+
+### Where to spill
+
+There are two options for us to spill: local disk or object store.
+
+1. Local Disk
+  - Pros: low latency, fixed costs, k8s could handle orphan spilled files in case of failure.
+  - Cons: limited capacity and throughput
+2. Object Store
+  - Pros: infinite capacity and throughput
+  - Cons: high latency, extra storage, and access cost, need to clean orphan spilled file.
+
+We can use a local disk first so that users don't need to pay any extra cost.
+
+### Spill directory
+
+Use need a pre-defined directory to hold all the temporary spilling files. In a cloud environment, k8s should be responsible for configuring this directory via the env variable `RW_BATCH_SPILL_DIR`. By default, we can use `/tmp/` as the default spill directory. Every executor should use a uuid to create a subdirectory under the default spill directory to ensure no conflict among them. After query processing, the executor are responsible for clearing the directory.
+
+
+### Serialization
+
+We can convert keys and values of the hash table into a chunk and then serialize the chunk into a protobuf bytes which has already been supported and used during shuffling between different tasks. Finally, spill file content will look like the below. As we can see this file write pattern is append-only and the read pattern is sequential scan. This can maximize the disk IO performance.
+
+Spill file format:
+
+```
+[proto_len]
+[proto_bytes]
+...
+[proto_len]
+[proto_bytes]
+
+```
```

Comment:
I am not very sure, should we add an CRC here? Personally, I tend to add CRC for any persisted data.

Reply:
Sounds good, we can add a CRC at the end of the file.